### PR TITLE
Support classExpressions in DTE

### PR DESCRIPTION
**Pull request by h-joo**
Except for the commit e9ca3d4d36924a880dda29ed3e1f0348f03ca092 everything else is non-functional, all others are fixing tests and moving things around, formatting.

### CODE REVIEW

File: src/compiler/transformers/declarations.ts [hunk 1]
```diff
@@ -331,7 +331,7 @@ export function transformDeclarations(context: TransformationContext) {
     let libs: Map<string, boolean>;
     let emittedImports: readonly AnyImportSyntax[] | undefined; // must be declared in container so it can be `undefined` while transformer's first pass
     const resolver = context.getEmitResolver();
-    const localInferenceResolver = createLocalInferenceResolver({
+    const { resolver: localInferenceResolver, isolatedDeclarations } = createLocalInferenceResolver({
```
Comment (dragomirtitian) [inline src/compiler/transformers/declarations.ts:334]: Like the discriminated union destructuring here.

File: src/compiler/transformers/declarations/localInferenceResolver.ts [hunk 2]
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
...
+                        currentSourceFile,
+                        {
+                            pos: node.heritageClauses[0].pos,
+                            end: node.heritageClauses[node.heritageClauses.length - 1].end,
+                        },
+                        Diagnostics.Heritage_clauses_in_class_expressions_are_not_allowed_with_isolatedDeclarations,
+                    ),
+                ],
+            });
+            invalid = true;
```
Comment (dragomirtitian) [inline src/compiler/transformers/declarations/localInferenceResolver.ts]: Really sad we have to not allow heritage clauses. 

I think we need to rephrase this a bit "not allowed" implies to me that you can't use them at all. You can, provided you add an annotation to the variable, we just can't infer from them.

File: src/compiler/transformers/declarations/localInferenceResolver.ts [hunk 3]
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
...
+                    member.parameters,
+                    type,
+                );
+                if (member.modifiers?.some(modifier => modifier.kind === SyntaxKind.StaticKeyword)) {
+                    staticMembers.push(methodSignature);
+                }
+                else {
+                    nonStaticMembers.push(methodSignature);
+                }
+            }
```
Comment (dragomirtitian) [inline src/compiler/transformers/declarations/localInferenceResolver.ts]: This seems very similar to inference for methods on object literals. Only it does not do inference for parameters with default values. Can we factor out the code in the object inference and reuse it here?

File: tests/baselines/reference/esDecorators-classDeclaration-setFunctionName(target=es2022).js [hunk 4]
```diff
@@ -62,20 +62,24 @@ let C = (() => {
...
+            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(null) : void 0;
+            __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
             C = _classThis = _classDescriptor.value;
+            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
             __runInitializers(_classThis, _classExtraInitializers);
         }
     };
     return C = _classThis;
 })();
+export default C;
```
Comment (dragomirtitian) [inline tests/baselines/reference/esDecorators-classDeclaration-setFunctionName(target=es2022).js:82]: Do we know why this changed ?

File: src/compiler/transformers/declarations/localInferenceResolver.ts [hunk 5]
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
...
+                    }
+                    nonStaticMembers.push(
+                        factory.createPropertySignature(
+                            [],
+                            member.name,
+                            /*questionToken*/ undefined,
+                            type,
+                        ),
+                    );
+                }
```
Comment (dragomirtitian) [inline src/compiler/transformers/declarations/localInferenceResolver.ts]: Similar to the object inference maybe we can refactor to a function. Also it should infer setter/getter type from the other accessor if present,

File: src/compiler/transformers/declarations/localInferenceResolver.ts [hunk 6]
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
...
+                if (member.modifiers?.some(modifier => modifier.kind === SyntaxKind.StaticKeyword)) {
+                    staticMembers.push(methodSignature);
+                }
+                else {
+                    nonStaticMembers.push(methodSignature);
+                }
+            }
+            else if (isGetAccessorDeclaration(member) || isSetAccessorDeclaration(member)) {
+                if (!hasGetSetAccessor) {
+                    hasGetSetAccessor = true;
```
Comment (dragomirtitian) [inline src/compiler/transformers/declarations/localInferenceResolver.ts]: Not sure I understand this. What if the class has multiple getter/setters for different names

File: src/compiler/transformers/declarations/localInferenceResolver.ts [hunk 7]
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
...
+                nonStaticMembers.push(member);
+            }
+            else if (isPropertyDeclaration(member)) {
+                const name = isPrivateIdentifier(member.name) ?
+                    // imitating the behavior from utilities.ts : getSymbolNameForPrivateIdentifier, but as we don't have
+                    // a Symbol & SymbolId in hand, we use NodeId of the declaration instead as an approximiation and to provide uniqueness.
+                    // TODO: This seems to have a high collision possibilitiy than the vanilla implementation as we have much less
+                    // ids for nodes in DTE.
+                    factory.createStringLiteral(`__#${node.parent.id}@${member.name.escapedText}`) :
+                    member.name;
```
Comment (dragomirtitian) [inline src/compiler/transformers/declarations/localInferenceResolver.ts:618]: I think the appropriate question here is *why* does TS keep private names in declarations at all. This feels like a TS bug. For class declarations the members are removed and `#private` is added to the class. There is no reason to expose what private members the class expression has and the type they have. 

```tsx
const x = class {
    #id = 10 // "__#3455@#id": number; in dedclarations 
}
class X {
    #id = 10 // removed and #private; added
}
```

[Playground Link](https://www.typescriptlang.org/play?ts=5.3.0-beta#code/MYewdgzgLgBAHjAvDYAbAhhCMDeBYAKBmJgGIBLAEyRgEYAGGAeiZgCIB9D0gZgBYArAIACFSmwBcMMAFcAtgCMApgCcA3DHJgYlJZTToV6KOXDZCAX0IGsMABq5CJMlRoNmrFUrkgAbnph0MGpSAAcVcl9jJQ10Sl1KSyA)

File: src/compiler/transformers/declarations/localInferenceResolver.ts [hunk 8]
```diff
@@ -340,162 +344,313 @@ export function createLocalInferenceResolver({
...
+                    name,
+                    member.questionToken,
+                    type,
+                );
+                if (member.modifiers?.some(modifier => modifier.kind === SyntaxKind.StaticKeyword)) {
+                    staticMembers.push(propertySignature);
+                }
+                else {
+                    nonStaticMembers.push(propertySignature);
+                }
```
Comment (dragomirtitian) [inline src/compiler/transformers/declarations/localInferenceResolver.ts]: Similar to Property assignment from object literals, again maybe we can reuse ?

**Review by dragomirtitian — CHANGES_REQUESTED** [review]

File: src/compiler/transformers/declarations/localInferenceResolver.ts (hunk 7 above)
Comment (h-joo) [reply src/compiler/transformers/declarations/localInferenceResolver.ts:618]: Okay, I'll put into the agenda for next week's meeting.

File: src/compiler/transformers/declarations/localInferenceResolver.ts (hunk 2 above)
Comment (h-joo) [reply src/compiler/transformers/declarations/localInferenceResolver.ts]: I made the message : "To use heritage clauses in class expressions with --isolatedDeclarations, you need explicit type annotation on the variable."

File: src/compiler/transformers/declarations/localInferenceResolver.ts (hunk 3 above)
Comment (h-joo) [reply src/compiler/transformers/declarations/localInferenceResolver.ts]: Done

File: src/compiler/transformers/declarations/localInferenceResolver.ts (hunk 8 above)
Comment (h-joo) [reply src/compiler/transformers/declarations/localInferenceResolver.ts]: Done.

File: src/compiler/transformers/declarations/localInferenceResolver.ts (hunk 6 above)
Comment (h-joo) [reply src/compiler/transformers/declarations/localInferenceResolver.ts]: you're right, I don't know what I was thinking while writing this code. I've adopted your code anyways so no longer relevant.

File: src/compiler/transformers/declarations/localInferenceResolver.ts (hunk 5 above)
Comment (h-joo) [reply src/compiler/transformers/declarations/localInferenceResolver.ts]: Done.

File: tests/baselines/reference/esDecorators-classDeclaration-setFunctionName(target=es2022).js (hunk 4 above)
Comment (h-joo) [reply tests/baselines/reference/esDecorators-classDeclaration-setFunctionName(target=es2022).js:82]: I have no idea, this seemed to be broken already in the target branch, but I don't have a good imagination to understand what would have caused this to change.

File: src/compiler/transformers/declarations/localInferenceResolver.ts [hunk 9]
```diff
@@ -388,7 +398,7 @@ export function createLocalInferenceResolver({
             }
             if (isComputedPropertyName(prop.name)) {
                 if (!resolver.isLiteralComputedName(prop.name)) {
-                    reportIsolatedDeclarationError(node);
+                    reportIsolatedDeclarationError(objectLiteral);
```
Comment (dragomirtitian) [inline src/compiler/transformers/declarations/localInferenceResolver.ts]: I think this is wrong, we should report on `prop.name` to help the user identify the issue. I know this was the case before, but could you fix it as long as we are making changes here?

File: src/compiler/transformers/declarations/localInferenceResolver.ts [hunk 10]
```diff
@@ -556,46 +583,26 @@ export function createLocalInferenceResolver({
                 }
             }
             else if (isMethodDeclaration(member)) {
-                const type = localInferenceFromInitializer(member, member.type);
-                if (!type) {
+                const { method, flags } = handleMethodDeclaration(member, member.name, NarrowBehavior.None);
+                if (flags === LocalTypeInfoFlags.Invalid) {
```
Comment (dragomirtitian) [inline src/compiler/transformers/declarations/localInferenceResolver.ts]: Could you please test with `&` for the invalid flag. Currently there is only one flag (used to be more) but I would like to keep the idea of this being a flag enum at least for now.

**Review by dragomirtitian — CHANGES_REQUESTED** [review]

File: src/compiler/transformers/declarations/localInferenceResolver.ts (hunk 9 above)
Comment (h-joo) [reply src/compiler/transformers/declarations/localInferenceResolver.ts]: Done

File: src/compiler/transformers/declarations/localInferenceResolver.ts (hunk 10 above)
Comment (h-joo) [reply src/compiler/transformers/declarations/localInferenceResolver.ts]: Done

**Pull request merged**